Index: configs/arcemu-optional.conf
===================================================================
--- configs/arcemu-optional.conf	(revision 2599)
+++ configs/arcemu-optional.conf	(working copy)
@@ -173,7 +173,8 @@
 	PvPKillAnnounce="0" 
 	Portals="0" 
 	L80etc="0" 
-	WeddingNPC="0">
+	WeddingNPC="0"
+	GMOnline="1">
 
 #-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#
 # GameMaster Configuration
Index: src/scripts/src/ExtraScripts/GMOnline.cpp
===================================================================
--- src/scripts/src/ExtraScripts/GMOnline.cpp	(revision 0)
+++ src/scripts/src/ExtraScripts/GMOnline.cpp	(revision 0)
@@ -0,0 +1,25 @@
+/*   
+	
+   Hungarian Half Scripting team (HHScripts).
+   Copyright (C) 2009  Twl
+
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.*/
+
+
+#include "StdAfx.h"
+#include "Setup.h"
+#pragma warning(disable:4305)		  //truncation int to char
+#pragma warning(disable:4309)
+void GmCheck(Player* pPlayer)
+{
+	if(pPlayer->GetSession()->CanUseCommand('az'))
+	{
+		char *message = new char[255];
+		sprintf(message, "%s[NOTE] %s<Admin>%s %s is available!", MSG_COLOR_YELLOW, MSG_COLOR_RED, MSG_COLOR_CYAN, pPlayer->GetName());
+		sWorld.SendWorldText(message);
+		delete message;
+	}
+	else if(pPlayer->GetSession()->CanUseCommand('a'))
+	{
+		char *messagew = new char[255];
+		sprintf(messagew, "%s[NOTE] %s<GM>%s %s is available!", MSG_COLOR_YELLOW, MSG_COLOR_RED, MSG_COLOR_CYAN, pPlayer->GetName());
+		sWorld.SendWorldText(messagew);
+		delete messagew;
+	}
+}
+
+void SetupGMOnline(ScriptMgr * mgr)
+{
+	mgr->register_hook(SERVER_HOOK_EVENT_ON_ENTER_WORLD, (void*)GmCheck);
+}
\ No newline at end of file
Index: src/scripts/src/ExtraScripts/Setup.cpp
===================================================================
--- src/scripts/src/ExtraScripts/Setup.cpp	(revision 2599)
+++ src/scripts/src/ExtraScripts/Setup.cpp	(working copy)
@@ -58,6 +58,13 @@
 	}
 */
 
+	//Sends a message to players if a GM has come online.
+	if(Config.OptionalConfig.GetBoolDefault("Extra", "GMOnline", false)
+	{
+		sLog.outColor(TGREEN,"\n	GMOnline script Enabled");
+		SetupGMOnline(mgr);
+	}
+
     // Allows users to type "+unstuck" without quotes to revive them if
 	// needed and teleport them to either Stormwind or Orgrimmar
     if(Config.OptionalConfig.GetBoolDefault("Extra", "Unstuck", false))
Index: src/scripts/src/ExtraScripts/Setup.h
===================================================================
--- src/scripts/src/ExtraScripts/Setup.h	(revision 2599)
+++ src/scripts/src/ExtraScripts/Setup.h	(working copy)
@@ -26,5 +26,6 @@
 void SetupPvPToken(ScriptMgr * mgr);
 //void SetupWeddingNPC(ScriptMgr * mgr);
 void SetupUnstuck(ScriptMgr * mgr);
+void SetupGMOnline(ScriptMgr * mgr);
 
 #endif
